docs/ldml/tr35-keyboards.md

```diff
@@ -1388,15 +1402,17 @@ _Attribute:_ `modifier` (required for `hardware`)
 > - `ctrlR`
 > - `shift`
 >
-> Note that `alt` is sometimes referred to as _opt_ or _option_.
+> Note that `alt` in this specification is referred to on some platforms as "opt" or "option".
 >
 > Left- and right- side modifiers (such as `"altL ctrlR"` or `"altL altR"`) should not be used together in a single `modifier` attribute value.
 >
-> Left- and right- side modifiers (such as `"altL ctrlR"` or `"altL altR"`) should not be used together in a single `modifier` attribute value.
-
+> Matching is done exactly. For example, if the `shift` and `altL` keys are held down, `<layer id="A" modifiers="altL"/>` will not match, but `<layer id="B" modifiers="altL shift"/>`will match.
```

Comment:
The spec has <layer id="layerId" modifier="{Set of Modifier Combinations}">
However, since it is a set, `modifiers` (with an S) would be the right attribute name.

There needs to be a precise, algorithmic definition of "match". That is, given an input set of modifiers IS, and a list of lines of the form:
<layer id="A" modifiers="{ModifierSet1}"/>
<layer id="A" modifiers="{ModifierSet2}"/>
...
It should be completely determinant which one of those lines matches. Some of the spec sounds like

Also, there need to be a precise list of exactly what the modifier key identifiers are. It appears to be the following, but it is not at all clear.

- none (no modifier, may not be combined with others)
- alt
- altL
- altR
- caps
- ctrl
- ctrlL
- ctrlR
- shift

If on the other hand, the IDs are purely up to the developer, then that opens the door to (massive) incompatibilities.

Phrases like "If there is a layer with a modifier alt, there may not be another layer with altL or altR. Similarly, if there is a layer with a modifier ctrl, there may not be a layer with ctrlL or ctrlR." are unclear. Does that mean such a keyboard is ill-formed? Must it be "rejected" — whatever that means?

```
Matching is done exactly. For example, if the shift and altL keys are held down, <layer id="A" modifiers="altL"/> will not match, but <layer id="B" modifiers="altL shift"/>will match.
``` 

This makes it sound like there must be an exact match, so if I have modifiers="shift", that cannot match an input set {shift, caps}, and so on. Is that really the case? If so, I can foresee a lot of duplicate layers in order to express current behavior of keyboards. If not, then the text of the specification is wrong.

Is the desire really to treat unmentioned modifiers as "a don't care", but pick the line with the most matches?

Anyway, this needs a precise specification.

Reply:
I'd like to take this into a separate PR, merging the text as-is, leaving this ticket open to amend the modifier logic.